### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by amishas157**
<details>
  <summary>PR Checklist</summary>

### PR Structure

- [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
- [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
      otherwise).
- [ ] This PR's title starts with the jira ticket associated with the PR.

### Thoroughness

- [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
- [ ] I've updated the README with the added features, breaking changes, new instructions on how to use the repository. I updated the description of the fuction with the changes that were made.

### Release planning

- [ ] I've decided if this PR requires a new major/minor/patch version accordingly to
    [semver](https://semver.org/), and I've changed the name of the BRANCH to release/_ , feature/_ or patch/\* .
</details>

### What

This PR adds integration tests for all the `export_*` commands. The changes comprise of following:
- Existing `runCLITest` in `cmd/export_ledgers_test.go` is used to run all test files in cmd.
- update=true flag is supported to update golden files automatically.
- More tests suites are added for export_ledger_entry_changes_test. This is to cover all kinds of ledger entry changes listed in https://github.com/stellar/stellar-etl?tab=readme-ov-file#unbounded-currently-unsupported
- Updated all the outdated golden files.
- A docker file is added to run tests. Current docker file does not have context of /cmd.
- A docker-compose file is added to start a docker compose service with right gcp creds and run the tests.
- Github workflow is added to run integration test, find code coverage and a threshold check for coverage
- Updated github workflows to cancel any previous jobs if new commits are pushed
- Updated Readme and Makefile for helper with running the tests.
- A secret `CREDS_TEST_HUBBLE` is added to this repo which has access to bigquery and storage bucket. This is used by integration test to read ledger backend data

### Why

To gain confidence in feature addition or bug fix, integration tests are required.

### Known limitations

None

### Test Results:
- Running integration test takes ~3 minutes.
Screenshot 2024-08-08 at 5 39 08 PM
- Our current test coverage is 57% and I have setup a threshold of 55%
Screenshot 2024-08-08 at 5 39 51 PM
- Commit intentionally failed to verify failure https://github.com/stellar/stellar-etl/pull/274/commits/1d39d476bb5027b3b205c134de18c58cc5cae9fe

### CODE REVIEW

**Comment by amishas157** [comment]
How I ran the tests:

- docker-compose build
- docker-compose run integration-tests go test -run ^TestExport ./cmd -timeout 30m -args -update=true

File: cmd/export_ledgers_test.go [hunk 1]
```diff
@@ -61,31 +57,25 @@ func TestExportLedger(t *testing.T) {
...
-			name:    "start too large",
-			args:    []string{"export_ledgers", "-s", "4294967295", "-e", "4294967295"},
-			golden:  "",
-			wantErr: fmt.Errorf("could not read ledgers: Latest sequence number is less than start sequence number (%d < 4294967295)", latestLedger),
-		},
-		{
-			name:    "end too large",
-			args:    []string{"export_ledgers", "-e", "4294967295", "-l", "4294967295"},
-			golden:  "",
-			wantErr: fmt.Errorf("could not read ledgers: Latest sequence number is less than end sequence number (%d < 4294967295)", latestLedger),
```
Comment (amishas157) [inline cmd/export_ledgers_test.go:76]: These are flaky test. Therefore I removed them

File: cmd/export_ledgers_test.go [hunk 2]
```diff
@@ -128,21 +118,37 @@ func sortByName(files []os.DirEntry) {
...
+					assert.NoError(t, err)
+				}
+			} else {
+				err = deleteLocalFiles(outLocation)
+				if err != nil {
+					log.Fatal(err)
+				}
+			}
+		}
+
```
Comment (amishas157) [inline cmd/export_ledgers_test.go:145]: This is to delete anything in the output directory before exporting any data.

File: cmd/export_ledger_entry_changes_test.go
Comment (sydneynotthecity) [inline cmd/export_ledger_entry_changes_test.go:1]: Do you think we'll run into any issues using different ledger ranges to generate these files?

File: cmd/export_ledgers_test.go [hunk 3]
```diff
@@ -61,31 +57,25 @@ func TestExportLedger(t *testing.T) {
...
-			golden:  "",
-			wantErr: fmt.Errorf("could not read ledgers: Latest sequence number is less than end sequence number (%d < 4294967295)", latestLedger),
+			wantErr: fmt.Errorf("Number of bytes written: 0"),
 		},
+		// {
+		// 	name:    "end too large",
+		// 	args:    []string{"export_ledgers", "-e", "4294967295", "-l", "4294967295"},
+		// 	golden:  "",
+		// 	wantErr: fmt.Errorf("could not read ledgers: Latest sequence number is less than end sequence number (%d < 4294967295)", latestLedger),
+		// },
```
Comment (sydneynotthecity) [inline cmd/export_ledgers_test.go:67]: I agree that this is a poorly designed test. Feel free to delete entirely

**Review by sydneynotthecity — COMMENTED** [review]
This looks great! 

I have a couple questions for my own understanding before approving. I also glanced over the golden set files briefly, but i'm assuming they are valid since the integration tests succeed.

File: .github/workflows/integration-tests.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      - name: Build and start services
+        run: |
+          docker-compose up -d
+
+      - name: Create GCP Credentials File
+        run: |
+          echo "$CREDS_TEST_HUBBLE" > ${{ runner.workspace }}/gcp-key.json
+        shell: bash
+        env:
+          CREDS_TEST_HUBBLE: ${{secrets.CREDS_TEST_HUBBLE}}
```
[the remark this answers is not in the document]
Comment (amishas157) [reply .github/workflows/integration-tests.yml:41]: No, since the secret is not logged anywhere. Even in the worst case secret is logged, github obsucures it with ***
https://stackoverflow.com/questions/59481933/how-can-i-extract-secrets-using-github-actions

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+      - name: Run tests
```
[the remark this answers is not in the document]
Comment (amishas157) [reply .github/workflows/integration-tests.yml:43]: Yes, CI is only meant for verification. Update should be done locally `make int-test-update`

File: cmd/export_contract_events_test.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"testing"
+)
+
+func TestExportContractEvents(t *testing.T) {
+	tests := []cliTest{
+		{
+			name:    "diagnostic events from one ledger",
+			args:    []string{"export_contract_events", "-s", "52271338", "-e", "52271350", "-o", gotTestDir(t, "one_ledger_txs.txt")},
```
[the remark this answers is not in the document]
Comment (amishas157) [reply cmd/export_contract_events_test.go:11]: > is there only one ledger written to the file? The export range is larger than 1

Is this because data is coming from test hubble?

> Also file name is a bit confusing if the output data is related to contract_events

Agree. Will rename it.

File: cmd/export_ledger_entry_changes_test.go
Comment (amishas157) [reply cmd/export_ledger_entry_changes_test.go:1]: That's a good point. There can be issues if code has bug for different kind of input. Since the ledger range can contain only certain variety of data. It may not cover for all test cases.

I had a thought to have custom fixtures along with ledger range test cases. But I think that's already part of our unit tests. so we should be good generally.

File: cmd/export_ledger_entry_changes_test.go [hunk 7]
```diff
@@ -24,16 +24,74 @@ func TestExportChanges(t *testing.T) {
...
+		},
+		{
+			name:              "Claimable balance from ledger entry",
+			args:              []string{"export_ledger_entry_changes", "-x", coreExecutablePath, "-c", coreConfigPath, "-s", "49265302", "-e", "49265400", "-o", gotTestDir(t, "claimable_balances/"), "--export-balances", "true"},
+			golden:            "claimable_balances.golden",
+			wantErr:           nil,
+			sortForComparison: true,
+		},
+		{
+			name:              "Trustllines from ledger entry",
```
Comment (amishas157) [inline cmd/export_ledger_entry_changes_test.go:48]:
```suggestion
			name:              "trustlines from ledger entry",
```

File: cmd/export_ledgers_test.go (hunk 3 above)
Comment (amishas157) [inline cmd/export_ledgers_test.go:67]:
```suggestion

```

File: .github/workflows/integration-tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (sydneynotthecity) [reply .github/workflows/integration-tests.yml:43]: nit: might be nice to add a section in the README that explains to the developer when they would need to update the golden files

File: cmd/export_ledger_entry_changes_test.go
Comment (sydneynotthecity) [reply cmd/export_ledger_entry_changes_test.go:1]: I think this is a good enough start point and if we run into edge cases, we can always either adjust the ledger ranges or add custom fixtures as you suggested

File: .github/workflows/integration-tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (amishas157) [reply .github/workflows/integration-tests.yml:43]: How about this? https://github.com/stellar/stellar-etl/pull/274/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R108-R110

**Review by sydneynotthecity — APPROVED** [review]
LGTM

File: .github/workflows/integration-tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (sydneynotthecity) [reply .github/workflows/integration-tests.yml:43]: works with me
